Lib/test/test_warnings/__init__.py

```diff
@@ -794,6 +794,36 @@ def test_stderr_none(self):
         self.assertNotIn(b'Warning!', stderr)
         self.assertNotIn(b'Error', stderr)
 
+    def test_issue31285(self):
+        # warn_explicit() should neither raise a SystemError nor cause an
+        # assertion failure, in case the return value of get_source() has a
+        # bad splitlines() method.
+        def _get_bad_loader(splitlines_ret_val):
+            class BadLoader:
+                def get_source(self, fullname):
+                    class BadSource(str):
+                        def splitlines(self):
+                            return splitlines_ret_val
+                    return BadSource('spam')
+            return BadLoader()
+        with support.captured_stderr() as stderr:
+            self.module.warn_explicit(
+                'foo', UserWarning, 'bar', 1,
+                module_globals={'__loader__': _get_bad_loader(42),
+                                '__name__': 'foobar'})
+        self.assertIn('bar:1: UserWarning: foo', stderr.getvalue())
+        show = self.module._showwarnmsg
+        try:
+            del self.module._showwarnmsg
+            with support.captured_stderr() as stderr:
+                self.module.warn_explicit(
+                    'foo', ArithmeticError, 'bar', 1,
```

Comment:
Is it valid to use ArithmeticError which is not a warning category?

Reply:
It is valid (IIUC, only `warning.warns()` requires the category to be a subclass of `Warning`). I used `ArithmeticError` because it caused `get_filter()` to return 'default'. But now it seems quite obscure to me, so I would change it.

Also, my test fails with `-Werror`, so I would fix that too.